### ISSUE TITLE
Cannot expand row on TreeGrid when first column has a NumberRenderer 

### ISSUE THREAD

**Issue by jrfinc** [vaadin/framework#9465]
If I set a NumberRenderer to a the first column of a TreeGrid, when I load the page, I see the arrow to expand the row, but when I click it it disappears entirely. It works OK without the NumberRenderer. 

I'm using v 8.1.0.beta1

**Comment by hesara** [comment]
I'm not able to reproduce this with current master. If this also happens for you with 8.1-SNAPSHOT, please provide a short example on how to reproduce this.

**Comment by jrfinc** [comment]
When I was trying to provide a short example, I realized this happens on a *very* particular case of ours, when I set the NumberRenderer inside a setStyleGenerator call for the column. The reason we do this, is because we have generic "grids" and then, depending on the class of the column (string, int) we style it differently. 

In particular, what I'm doing is: 

`Iterator<Grid.Column> it = grid.getColumns().iterator();
        while (it.hasNext()) {
            Grid.Column c = it.next();
            c.setStyleGenerator(obj -> {
                Object value = c.getValueProvider().apply(obj);
                if (c.getId().equals("id") && !(value instanceof String)) {
                    c.setRenderer(new NumberRenderer(df));
                    return "align-right";
                }
                return null;
            });
        }`

I could set the renderer outside, but in that case, I wouldn't know the type of the cell (It is based in this link: https://vaadin.com/forum#!/thread/15554849). 

Here is a full example: https://gist.github.com/jrfinc/29975d3c8320425ce8aa81ff5c18c33f

This works OK in a normal Grid. but on a TreeGrid it breaks. 
I tried it with the 8.1.SNAPSHOT also. 

I understand this is very particular, so feel free to close this issue if this is not the place to discuss this. 

Thanks.

**Comment by tsuoanttila** [comment]
From a quick look at the code example, your style generator is resetting the renderer of the id column quite often. the style generator is called for each object being sent to the client-side. You should set the renderer only once (hopefully during the init of the column), and if this style generator way is the only one that works for you, then you should actually check if the renderer is already an instanceof number renderer. 

There's also a known issue when you change the renderer of the hierarchy column (that's why it should throw an exception), but that might not be the case here (unless you see an exception when running that code).

Changing the renderer of the hierarchy column causes it to lose the hierarchy renderer wrapping, and this makes the arrow disappear.

**Comment by jrfinc** [comment]
Thanks for your suggestions. Yes, I understand setting the renderer here is far from ideal, but haven't found another solution yet to getting the class of the column if it's not when setting the styleGenerator.  

As to checking if the renderer is already of type NumberRenderer, the column no longer has a getRenderer (it used to have one on v7), but it was left out on v8.  

Also, I am indeed setting the renderer on the Hierarchy column but no exception is being thrown. 

Thanks again. I guess for now I'll have to use that column without the renderer it used to have on v7.

**Comment by tsuoanttila** [comment]
I took a look at this, and created a change that actually allows the change of renderer for the hierarchy column. But there's another issue related to this, which is the issues with collapsing nodes when data gets reset for any reason (one example ticket #9449, fix pending #9496).

I'll push in a fix and a test to verify that the hierarchy column renderer can be changed and it will still work after that. Added the `Column#getRenderer` while I was at it.

**Comment by tsuoanttila** [comment]
So the remaining problem after that pull requests gets merged, is that the renderer switch in treegrid causes a full data reset (to update all items) which collapses all expanded nodes (including the one that was opened moment before)

**Comment by jrfinc** [comment]
Thanks for your time @tsuoanttila

**Comment by tsuoanttila** [comment]
There should be a build of 8.1-SNAPSHOT running now (available in about an hour) that should contain the `Column#getRenderer` method that should reduce the full resets you need to make and possibly even solve part of this problem. You should add a check of the renderer type and not re-set it if it's already ok.

**Comment by jrfinc** [comment]
Great thanks again @tsuoanttila!
